### DISCUSSION TITLE
TypeVar against Any, but needs to propagate correctly in the face of Union w/ another sub-type

### DISCUSSION THREAD

**Discussion by zzzeek** [python/typing#1143] [Q&A]
This one happens on Mypy but not on Pyright.    In the example below, `function_two()` would like to return either an instance of `_T` or an integer status code.  however trying to call upon `function_two()`, mypy expands out `_T` to include the type of the status code as well.  It doesnt do this in the identical case where status code would be the None constant.  Seems kind of buggy but at the same time I can sort of guess how it is making this mistake, so not sure how to proceed other than returning Any for `function_two()` or otherwise degrading the quality of typing.    Making _T a more specific TypeVar fixes, but the use case is against any arbitrary user-defined class.  I would, if possible, have it exclude integers, but I dont know how to make a TypeVar that excludes certain types only.

```py
from __future__ import annotations


from typing import Any
from typing import Generic
from typing import TypeVar
from typing import Union

_T = TypeVar("_T", bound=Any)


class Thing(Generic[_T]):
    pass


def caller_one(
    thing: Thing[_T],
) -> Union[_T, None]:

    return function_one(thing)


def function_one(
    thing: Thing[_T],
) -> Union[_T, None]:
    ...


def caller_two(
    thing: Thing[_T],
) -> Union[_T, int]:

    return function_two(thing)


def function_two(
    thing: Thing[_T],
) -> Union[_T, int]:
    ...
```

output:

```
$ mypy test3.py 
test3.py:33: error: Argument 1 to "function_two" has incompatible type "Thing[_T]"; expected "Thing[Union[_T, int]]"
Found 1 error in 1 file (checked 1 source file)
```

**Comment by zzzeek** [comment]
going to try this:

```py
def caller_two(
    thing: Thing[_T],
) -> Union[_T, int]:

    return cast(Union[_T, int], function_two(thing))

``` 

that seems to convince it

**Reply by erictraut** [reply]
Looks to be an issue with bidirectional type inference. If you assign the value to an intermediate value before returning it, that seems to fix it.

```python
def caller_two(
    thing: Thing[_T],
) -> Union[_T, int]:
    temp = function_two(thing)
    return temp
```

**Reply by zzzeek** [reply]
weird, ill try that too thanks!
